### PR TITLE
Similar dev search low connectivity based

### PR DESCRIPTION
**Pull request by TheTriangle**
For this task, we gather candidates from stargazers of stargazed repositories. For each candidate, we independently gather their variable names and languages from their repositories, and then calculate similarities with the starting developer

### CODE REVIEW

File: DOCKERFILE [hunk 1]
```diff
@@ -0,0 +1,23 @@
+FROM python:3.10.6
```
Comment (EgorBu) [inline DOCKERFILE:1]: move it to separate pr - and let's call this file `Dockerfile` - so github will automatically apply correct highlighting

File: src/main.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+async def get_candidates(start_developer: Developer, asyncio_client: httpx.AsyncClient = None) -> List[Developer]:
+    """
+    Return list of developers (candidates) from stargazers of stargazed repositories
+    :param start_developer: developer, from whom to start looking for candidates
+    :param asyncio_client: asyncio client to perform requests from
+    :return:
+    """
+    developer_stargazes = await start_developer.get_stargazes(asyncio_client)
+    tasks = []
+    candidates = set[Developer]()
```
Comment (EgorBu) [inline src/main.py:21]: nice

File: src/main.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
+import asyncio
```
Comment (EgorBu) [inline src/main.py:1]: let's rename `src` to some meaningful module name
If you want to make module callable like `python -m your_module --arg1 arg1 --arg2 arg2 ...` - you can rename `main.py` to `__main__.py`

File: src/main.py [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    """
+    result = []
+    i = 0
+    for developer in tqdm(developers, desc='computing similarities between developers'):
+        result.append(asyncio.run(main_developer.compute_similarity(developer)))
+    return result
+
+
+if __name__ == '__main__':
+    starting_developer_url = input('Enter the github url of the developer, for which to find similar devs: ')
```
Comment (EgorBu) [inline src/main.py]: let's use Click to make proper Command Line Interface - just thin wrapper around this logic

Comment (EgorBu) [reply src/main.py]: and let's make an argument number of similar developers to print

File: src/main.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
+import asyncio
+from typing import List
+
+import httpx
+
+from model.constants import max_candidates_num
+from model.developer import Developer
+from model.fetcher import fetch_stargazers_for_repo
+from model.repository import Repository
+from tqdm import tqdm
```
Comment (EgorBu) [inline src/main.py]: Let's follow import order suggested by pep8
* std lib
* external libs
* imports from the current module

File: src/main.py [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+    if asyncio_client is None:
+        await client.aclose()
+
+    for page_candidates in response:
+        candidates.update([Developer(candidate_url) for candidate_url in page_candidates])
+
+    return list(candidates)
+
+
+def compute_similarities(main_developer, developers) -> List[float]:
```
Comment (EgorBu) [inline src/main.py:43]: typing is missed for arguments

File: src/main.py [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+from model.repository import Repository
+from tqdm import tqdm
+
+
+async def get_candidates(start_developer: Developer, asyncio_client: httpx.AsyncClient = None) -> List[Developer]:
+    """
+    Return list of developers (candidates) from stargazers of stargazed repositories
+    :param start_developer: developer, from whom to start looking for candidates
+    :param asyncio_client: asyncio client to perform requests from
+    :return:
```
Comment (EgorBu) [inline src/main.py:17]: let's describe return

File: src/main.py [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+    """
+    developer_stargazes = await start_developer.get_stargazes(asyncio_client)
```
Comment (EgorBu) [inline src/main.py:19]: typo in name `get_stargazes` - `get_stargazers`

File: src/main.py [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+from model.developer import Developer
+from model.fetcher import fetch_stargazers_for_repo
+from model.repository import Repository
+from tqdm import tqdm
+
+
+async def get_candidates(start_developer: Developer, asyncio_client: httpx.AsyncClient = None) -> List[Developer]:
+    """
+    Return list of developers (candidates) from stargazers of stargazed repositories
+    :param start_developer: developer, from whom to start looking for candidates
```
Comment (EgorBu) [inline src/main.py:15]: it should be a start repository to find stargazers of repository

File: src/main.py [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+    result = []
+    i = 0
+    for developer in tqdm(developers, desc='computing similarities between developers'):
+        result.append(asyncio.run(main_developer.compute_similarity(developer)))
+    return result
+
+
+if __name__ == '__main__':
+    starting_developer_url = input('Enter the github url of the developer, for which to find similar devs: ')
+    starting_developer = Developer(starting_developer_url)
```
Comment (EgorBu) [inline src/main.py:60]: starting repository -> list of stargazers -> list of top repositories starred by stargazers of start repository

File: src/model/constants.py [hunk 11]
```diff
@@ -0,0 +1,9 @@
+headers = {'Authorization': 'token '}
+
+token = input('Github token: ')
```
Comment (EgorBu) [inline src/model/constants.py]: provide a name of environment variable to read github token from - it's more secure way to do it. Anyway - it should be part of CLI instead of asking it during import
If you are creating constants - let's follow pep8 naming convention for constants - it should be UPPER_CASE

Comment (EgorBu) [reply src/model/constants.py]: and (it looks like) all these constants could be default values to CLI

**Review by EgorBu — CHANGES_REQUESTED** [review]
Code looks nice! Let's polish this part and move further

File: src/model/constants.py (hunk 11 above)
Comment (TheTriangle) [reply src/model/constants.py]: about setting token in environment variable - i'm not sure it's the best solution, and my colleagues seem to agree. It's supposed to be generated for one-time use, that's what should make it secure
